File: examples/ts/mcp-custom-agent-flow.ts
import { AgenticExecutor, DisperslClient } from "@codefundi/dispersl-sdk";

const client = new DisperslClient({
  baseUrl: process.env.DISPERSL_API_URL ?? "https://api.dispersl.com/v1",
  apiKey: process.env.DISPERSL_API_KEY ?? ""
});

async function main() {
  const executor = new AgenticExecutor(client, async (tool) => {
    if (tool.function.name === "get_release_calendar") {
      return {
        toolName: "get_release_calendar",
        status: "success" as const,
        output: JSON.stringify({ freezeDate: "2026-04-15", releaseDate: "2026-04-22" })
      };
    }
    return { toolName: tool.function.name, status: "success" as const, output: tool.function.arguments };
  });

  executor.mcpTools.register({
    name: "get_release_calendar",
    description: "Return release calendar dates for planning and rollout.",
    parameters: {
      type: "object",
      properties: {},
      required: []
    },
    execute: async () => ({ freezeDate: "2026-04-15", releaseDate: "2026-04-22" })
  });

  const planned = await executor.runPlanAndAgentLoop({
    prompt: "Create a release plan for onboarding automation in regulated environments.",
    agentChoices: "auto"
  });
  console.log("Planned task", planned.taskId);

  const completed = await executor.runAgentCompletionLoop({
    nameId: "release-manager",
    prompt: "Finalize the release checklist using available custom tools."
  });
  console.log("Completion events", completed.events.length);
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});

File: examples/py/mcp_custom_agent_flow.py
import asyncio
import json

from dispersl import AgenticExecutor, AsyncDisperslClient, MCPTool, ToolResult


async def tool_runner(tool: dict) -> ToolResult:
    name = tool["function"]["name"]
    if name == "get_release_calendar":
        output = json.dumps({"freeze_date": "2026-04-15", "release_date": "2026-04-22"})
        return ToolResult(tool_name=name, status="success", output=output)
    return ToolResult(tool_name=name, status="success", output=tool["function"]["arguments"])


async def main() -> None:
    client = AsyncDisperslClient(
        base_url="https://api.dispersl.com/v1",
        api_key="YOUR_API_KEY",
    )
    executor = AgenticExecutor(client, tool_runner)

    executor.mcp_tools.register(
        MCPTool(
            name="get_release_calendar",
            description="Return release calendar dates for planning and rollout.",
            parameters={"type": "object", "properties": {}, "required": []},
            execute=lambda _args: {"freeze_date": "2026-04-15", "release_date": "2026-04-22"},
        )
    )

    # Plan does not require MCP tools. Use auto to discover the right team.
    planned = await executor.run_plan_and_agent_loop(
        prompt="Create a release plan for onboarding automation in regulated environments.",
        agent_choices="auto",
    )
    print("Planned task:", planned["task_id"])

    # Completion run can use MCP-backed custom tools with a specific custom agent.
    completed = await executor.run_agent_completion_loop(
        name_id="release-manager",
        prompt="Finalize release checklist with custom tool data.",
    )
    print("Completion events:", len(completed["events"]))
    await client.aclose()


if __name__ == "__main__":
    asyncio.run(main())
